plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/utils/KubernetesClusterUtil.java

```diff
@@ -0,0 +1,306 @@
+// specific language governing permissions and limitations
+// under the License.
+
+package com.cloud.kubernetes.cluster.utils;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.net.URL;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.log4j.Logger;
+
+import com.cloud.kubernetes.cluster.KubernetesCluster;
+import com.cloud.uservm.UserVm;
+import com.cloud.utils.Pair;
+import com.cloud.utils.StringUtils;
+import com.cloud.utils.ssh.SshHelper;
+import com.google.common.base.Strings;
+
+public class KubernetesClusterUtil {
+
+    protected static final Logger LOGGER = Logger.getLogger(KubernetesClusterUtil.class);
+
+    public static boolean isKubernetesClusterNodeReady(KubernetesCluster kubernetesCluster, String ipAddress, int port,
+                                                       String user, File sshKeyFile, String nodeName) throws Exception {
+        Pair<Boolean, String> result = SshHelper.sshExecute(ipAddress, port,
+                user, sshKeyFile, null,
+                String.format("sudo kubectl get nodes | awk '{if ($1 == \"%s\" && $2 == \"Ready\") print $1}'", nodeName.toLowerCase()),
+                10000, 10000, 20000);
+        if (result.first() && nodeName.equals(result.second().trim())) {
+            return true;
+        }
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug(String.format("Failed to retrieve status for node: %s in Kubernetes cluster ID: %s. Output: %s", nodeName, kubernetesCluster.getUuid(), result.second()));
+        }
+        return false;
+    }
+
+    public static boolean isKubernetesClusterNodeReady(KubernetesCluster kubernetesCluster, String ipAddress, int port,
+                                                       String user, File sshKeyFile, String nodeName, int retries,
+                                                       int waitDuration) {
+        int retryCounter = 0;
+        while (retryCounter < retries) {
+            boolean ready = false;
+            try {
+                ready = isKubernetesClusterNodeReady(kubernetesCluster, ipAddress, port, user, sshKeyFile, nodeName);
+            } catch (Exception e) {
+                LOGGER.warn(String.format("Failed to retrieve state of node: %s in Kubernetes cluster ID: %s", nodeName, kubernetesCluster.getUuid()), e);
+            }
+            if (ready) {
+                return true;
+            }
+            try {
+                Thread.sleep(waitDuration);
+            } catch (InterruptedException ie) {
+                LOGGER.error(String.format("Error while waiting for Kubernetes cluster ID: %s node: %s to become ready", kubernetesCluster.getUuid(), nodeName), ie);
+            }
+            retryCounter++;
+        }
+        return false;
+    }
+
+    public static boolean uncordonKubernetesClusterNode(KubernetesCluster kubernetesCluster, String ipAddress, int port,
```

Comment:
uncordon? can you improve the name or add a good javadoc? as this is a worker utility, i don't see this a as a big issue, but for the non initiated it is a daunting name.

Reply:
@DaanHoogland Named so because that is the Kubernetes action performed here, https://github.com/apache/cloudstack/pull/3680/files/b46dcc6cde3441010f14dd9c0f59910db00ab966#diff-0ed1fd9dd0afbdce8420ae7b5f2ec732R89-R91
Should I add javadoc?

Reply:
public method so yes, wouldn't hurt

Reply:
Added javadoc